packages/core/src/driver/Playwright.ts

```
@@ -0,0 +1,184 @@
+	close(): Promise<void>
+	reopenPage(incognito: boolean): Promise<void>
+	closePages(): Promise<void>
+	isFirstRun(): boolean
+}
+
+export class PlaywrightClient implements PlaywrightClientLike {
+	constructor(public browser: Browser, public page: Page) {}
+
+	private _isClosed = false
+	private _firstRun = true
+
+	async close(): Promise<void> {
+		if (this._isClosed) return
+		await this.browser.close()
+		this._isClosed = true
+	}
+
+	isFirstRun(): boolean {
+		return this._firstRun
+	}
+
+	async reopenPage(incognito = false): Promise<void> {
+		if (this._firstRun) {
+			this._firstRun = false
+			return
+		}
+		await this.closePages()
+		if (incognito) {
+			const context = await this.browser.newContext()
+			this.page = await context.newPage()
+		} else {
+			this.page = await this.browser.newPage()
+		}
+	}
+
+	async closePages(): Promise<void> {
+		const pages = this.page.context().pages()
+		for (const page of pages) {
+			await page.close()
+		}
+	}
+}
+
+export async function launchBrowserServer(
+	passedOptions: Partial<ConcreteLaunchOptions> = {},
+): Promise<BrowserServer> {
+	const options: ConcreteLaunchOptions = {
+		...defaultLaunchOptions,
+		...passedOptions,
+	}
+
+	if (!options.sandbox && ['chrome', 'chromium'].includes(options.browser)) {
+		options.args.push('--no-sandbox')
+	}
+
+	if (options.debug) {
+		console.dir(options)
+	}
+
+	if (options.browser !== 'webkit') {
+		options.args.push('--auth-server-whitelist="hostname/domain"')
+	}
+
+	let browserType = options.browser || 'chromium'
+	// can not run multiple users on chrome, need to switch back to chromium
+	if (browserType === 'chrome') {
+		browserType = 'chromium'
+	}
+	return playwright[browserType].launchServer(options)
+}
+
+export async function connectWS(wsEndpoint: string, type?: BrowserType): Promise<PlaywrightClient> {
+	let browserType = type || 'chromium'
+	// can not run multiple users on chrome, need to switch back to chromium
+	if (browserType === 'chrome') {
```

This seems like a bad idea, we need to support Chrome stable to get access to proprietary codecs for supporting video playback, if we always force it back to chromium this will silently fail.

---

I'm not sure about the video playback here, this function is using for running multiple users locally. In the past, we already supported 3 kinds which bundled in playwright are `chromium, firefox, webkit`. We just add chrome stable into this list a few days ago but we have an issue with chrome https://github.com/microsoft/playwright/issues/5327. So I think we should keep supporting 3 kinds of bundled browsers of Playwright.

---

This looks like a breaking change, I mean, I can't see how this would work in production, single or multiple user mode

---

Hi Ivan, this feature support running multiple users locally, not in production. we can trigger this feature by the setting:
```
export const settings: TestSettings = {
  stages: [
    {
      duration: '10m',
      target: 3
    }
  ]
}
```
we have a `scheduler` package to do this thing. Firstly we need to start a browser server, then it depends on the number of targets, we connect to the `wsEndpoint` to init the client browser.
So this feature only runs locally. we ignore this setting when running in production.

---

Okay, I see. Cool.